File: LLD NodeJs/p1_oops/c7_composition/composition/Car.js
const Engine = require('./Engine');

/**
 * Car class demonstrating Composition (strong HAS-A relationship)
 * 
 * Composition: Car HAS-A Engine, Engine CANNOT exist without Car.
 * The Engine is created INSIDE the Car, not passed from outside.
 */
class Car {
    #model;
    #engine; // Engine is created INSIDE Car (strong ownership)

    constructor(model, engineType) {
        this.#model = model;
        console.log("Creating Car: " + model);
        this.#engine = new Engine(engineType); // Engine is created HERE
    }

    drive() {
        console.log("Car " + this.#model + " is driving.");
        this.#engine.start();
    }

    park() {
        console.log("Car " + this.#model + " is parked.");
        this.#engine.stop();
    }

    // When Car is destroyed (garbage collected), Engine is also destroyed
    // Engine cannot exist independently
    // This is COMPOSITION (strong relationship)
}

module.exports = Car;

File: LLD/src/main/java/org/example/p1_oops/c7_composition/composition/Car.java
package org.example.p1_oops.c7_composition.composition;

// Composition: Car HAS-A Engine, Engine CANNOT exist without Car
public class Car {
    private String model;
    private Engine engine; // Engine is created INSIDE Car (strong ownership)

    public Car(String model, String engineType) {
        this.model = model;
        System.out.println("Creating Car: " + model);
        this.engine = new Engine(engineType); // Engine is created HERE
    }

    public void drive() {
        System.out.println("Car " + model + " is driving.");
        engine.start();
    }

    public void park() {
        System.out.println("Car " + model + " is parked.");
        engine.stop();
    }

    // When Car is destroyed, Engine is also destroyed
    // Engine cannot exist independently
    // This is COMPOSITION (strong relationship)
}
